### ISSUE TITLE
GPU support in PyTorch XLA

### ISSUE THREAD

**Issue by goswamig** [pytorch/xla#2272]
## ❓ Questions and Help

Do we have GPU support in PyTorch XLA? If so would you please point me to any resource to begin with ?

**Comment by jysohn23** [comment]
Yes, it does support GPUs, but we have limited testing with them tbh. We've just set up a release pipeline for `torch_xla` wheels built with GPU support, so we'll work on testing them more internally, adding nightly tests, and create a sample PyTorch/XLA GPU notebook from it once it's ready.

**Comment by gautamkmr** [comment]
I am happy to help to test it out, if you have any doc to setup and run basic MNIST ?

**Comment by dlibenzi** [comment]
Maybe check here: https://github.com/pytorch/xla/issues/2121

**Comment by gautamkmr** [comment]
Thank you @dlibenzi  I tried those steps, container build was successful but then `import torch_xla` was failing. I think we can polish that instruction. 
Also do you have wheel or docker container build which I can test or use ?

**Comment by zcain117** [comment]
Sorry for the delay @gautamkmr , we are getting close to something that we can share but we're not quite there yet. We're planning on offering wheels and a docker image that you can use just like our normal wheels/docker image.

Once that is ready, it should be really easy to switch between TPU and GPU without having to change your code. We can let you know once it's ready if you're still interested in trying it

**Comment by gupta-abhay** [comment]
@zcain117 - I am interested in trying this as well. Please let me know so that I can give this a shot.  Thank you!

**Comment by tyoc213** [comment]
Will I be able to install on a normal PC with CUDA support?

**Comment by zcain117** [comment]
@goswamig @gupta-abhay @tyoc213 

I tried this Friday and it worked for me:
1. Made a VM in Google Cloud using the boot disk: "Deep Learning Image: Pytorch 1.4 and fastai m55". Sub-description: "PyTorch 1.4.0 (and fastai) with CUDA 10.0 and Intel® MKL-DNN, Intel® MKL." I think any VM that has up-to-date CUDA installed would probably work.
2. Pulled the pytorch/xla docker image: `docker run --gpus all -it --shm-size 16G gcr.io/tpu-pytorch/xla@sha256:efe47b7a3875ddfa3ea9c68a12ed5517c19cbb3cf272776fba64bec8a683299f`. This was Friday's pytorch/xla CUDA image but the latest nightly should also work. The tag would be `nightly_3.6_cuda` or `nightly_3.7_cuda` or you can use a dated image like `nightly_3.6_cuda_20200914`.
3. Ran training with a new environment variable: `GPU_NUM_DEVICES=4 python test/test_train_mp_mnist.py`

I was able to train on GPUs with 0 code change. We don't have automated tests for XLA GPU so we won't call this officially supported yet, but I'm working on making those tests.

Let me know if you're able to get it working. If you run into any issues, please open a new issue that has the error you're seeing and details about VM software, TPU software, CUDA version, etc.

**Comment by tyoc213** [comment]
@zcain117 will work on my PC? (will try with latest nighly now)

```
$ docker run --gpus all -it --shm-size 16G gcr.io/tpu-pytorch/xla@sha256:efe47b7a3875ddfa3ea9c68a12ed5517c19cbb3cf272776fba64bec8a683299f
docker: Error response from daemon: could not select device driver "" with capabilities: [[gpu]].
ERRO[0000] error waiting for container: context canceled 
```

after install

```
a$ docker run --gpus all -it --shm-size 16G gcr.io/tpu-pytorch/xla@sha256:efe47b7a3875ddfa3ea9c68a12ed5517c19cbb3cf272776fba64bec8a683299f
Unable to find image 'gcr.io/tpu-pytorch/xla@sha256:efe47b7a3875ddfa3ea9c68a12ed5517c19cbb3cf272776fba64bec8a683299f' locally
sha256:efe47b7a3875ddfa3ea9c68a12ed5517c19cbb3cf272776fba64bec8a683299f: Pulling from tpu-pytorch/xla
f08d8e2a3ba1: Pull complete 
3baa9cb2483b: Pull complete 
94e5ff4c0b15: Pull complete 
1860925334f9: Pull complete 
05cc64cc481f: Pull complete 
184e72acd9ad: Pull complete 
d74ed5c8861b: Pull complete 
c9b52bf7e9d8: Pull complete 
7dcf5949bbe9: Pull complete 
4d3546b091ff: Pull complete 
098c28111250: Pull complete 
8f8a577c6d35: Pull complete 
14ab1e0400c8: Pull complete 
2856e33662ea: Pull complete 
0b76d9d0d825: Pull complete 
91c0256fead5: Pull complete 
bb161fce3332: Pull complete 
c61fc1e53651: Pull complete 
b3a7fe53500b: Pull complete 
Digest: sha256:efe47b7a3875ddfa3ea9c68a12ed5517c19cbb3cf272776fba64bec8a683299f
Status: Downloaded newer image for gcr.io/tpu-pytorch/xla@sha256:efe47b7a3875ddfa3ea9c68a12ed5517c19cbb3cf272776fba64bec8a683299f
docker: Error response from daemon: could not select device driver "" with capabilities: [[gpu]].
ERRO[6920] error waiting for container: context canceled 
```
